From f52525338c9c4e89b7001e685b208538b68f441d Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@freescale.com>
Date: Wed, 24 Aug 2011 15:00:49 +0800
Subject: [PATCH 654/670] ENGR00144145-1 usb: can be built as loadable modules

MSL part

Follow below load sequence to use OTG

- fsl_otg_arc
- ehci-hcd
- arcotg_udc

Besides, in order to match i.mx53 release windows, it changes some
common code, there will be a better solution at 2.6.38 in future

Signed-off-by: Peter Chen <peter.chen@freescale.com>
---
 arch/arm/mach-mx5/usb_dr.c     | 2 +-
 arch/arm/mach-mx5/usb_h1.c     | 9 +--------
 arch/arm/mach-mx5/usb_h2.c     | 8 +-------
 arch/arm/plat-mxc/usb_common.c | 8 +++++++-
 4 files changed, 10 insertions(+), 17 deletions(-)

diff --git a/arch/arm/mach-mx5/usb_dr.c b/arch/arm/mach-mx5/usb_dr.c
index 98637be..c4caf07 100644
--- a/arch/arm/mach-mx5/usb_dr.c
+++ b/arch/arm/mach-mx5/usb_dr.c
@@ -224,7 +224,7 @@ static void host_wakeup_handler(struct fsl_usb2_platform_data *pdata)
 {
 	_host_wakeup_enable(pdata, false);
 	_host_phy_lowpower_suspend(pdata, false);
-	fsl_usb_recover_hcd(&mxc_usbdr_host_device);
+	pdata->wakeup_event = 1;
 }
 /* End of host related operation for DR port */
 #endif /* CONFIG_USB_EHCI_ARC_OTG */
diff --git a/arch/arm/mach-mx5/usb_h1.c b/arch/arm/mach-mx5/usb_h1.c
index 16b7cfa..9e448ed 100644
--- a/arch/arm/mach-mx5/usb_h1.c
+++ b/arch/arm/mach-mx5/usb_h1.c
@@ -28,13 +28,6 @@ static struct clk *usb_phy2_clk;
 static struct clk *usb_oh3_clk;
 static struct clk *usb_ahb_clk;
 extern int clk_get_usecount(struct clk *clk);
-
-#ifdef CONFIG_USB_EHCI_ARC
-extern void fsl_usb_recover_hcd(struct platform_device *pdev);
-#else
-static void fsl_usb_recover_hcd(struct platform_device *pdev)
-{; }
-#endif
 /*
  * USB Host1 HS port
  */
@@ -127,7 +120,7 @@ static void h1_wakeup_handler(struct fsl_usb2_platform_data *pdata)
 {
 	_wake_up_enable(pdata, false);
 	_phy_lowpower_suspend(pdata, false);
-	fsl_usb_recover_hcd(&mxc_usbh1_device);
+	pdata->wakeup_event = 1;
 }
 
 static void usbh1_wakeup_event_clear(void)
diff --git a/arch/arm/mach-mx5/usb_h2.c b/arch/arm/mach-mx5/usb_h2.c
index a4a3f95..e2139c4 100644
--- a/arch/arm/mach-mx5/usb_h2.c
+++ b/arch/arm/mach-mx5/usb_h2.c
@@ -22,12 +22,6 @@
 #include "iomux.h"
 #include "mx51_pins.h"
 
-#ifdef CONFIG_USB_EHCI_ARC
-extern void fsl_usb_recover_hcd(struct platform_device *pdev);
-#else
-static void fsl_usb_recover_hcd(struct platform_device *pdev)
-{; }
-#endif
 /*
  * USB Host2 HS port
  */
@@ -109,7 +103,7 @@ static void h2_wakeup_handler(struct fsl_usb2_platform_data *pdata)
 {
 	_wake_up_enable(pdata, false);
 	_phy_lowpower_suspend(pdata, false);
-	fsl_usb_recover_hcd(&mxc_usbh2_device);
+	pdata->wakeup_event = 1;
 }
 
 static void usbh2_wakeup_event_clear(void)
diff --git a/arch/arm/plat-mxc/usb_common.c b/arch/arm/plat-mxc/usb_common.c
index 573c159..5164172 100644
--- a/arch/arm/plat-mxc/usb_common.c
+++ b/arch/arm/plat-mxc/usb_common.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2004-2010 Freescale Semiconductor, Inc. All Rights Reserved.
+ * Copyright (C) 2004-2013 Freescale Semiconductor, Inc. All Rights Reserved.
  */
 
 /*
@@ -879,6 +879,12 @@ void usbotg_uninit(struct fsl_usb2_platform_data *pdata)
 		if (pdata->xcvr_type == PORTSC_PTS_SERIAL)
 			clk_disable(usb_clk);
 		clk_disable(usb_ahb_clk);
+	/* FIXME workaround for clock refcounf when load/unload usb module
+	 * Need to fix it in future
+	 */
+	#ifdef CONFIG_USB_OTG
+		clk_disable(usb_ahb_clk);
+	#endif
 	}
 }
 EXPORT_SYMBOL(usbotg_uninit);
-- 
1.8.0

